enhancements

thiswillbeyourgithub opened the pull request:
- feat: added progress bar
- feat: added exclusion list
- feat: asks twice for password
- docs: mentions exclusion list
- docs: typo

I hope you don't mind me creating a PR!

Note that I have no training in cryptography or security and intends solely to slightly improve your project

Commit:
feat: added progress bar

Commit:
feat: added exclusion list

Commit:
feat: asks twice for password

Commit:
docs: mentions exclusion list

Commit:
docs: typo

Commit:
style: matches the original author's coding style

Commit:
feat: implements local backup

Commit:
feat: upload smaller file first

Commit:
minor: phrasing

Commit:
fix: adds check for unix special file and others

Commit:
added flag to indicate wether to treat larger files first or last

Commit:
modified:   nfreezer.py
        progress bar is in bytes now
        but I had to change the indentation so git thinks a lot of lines changes

Commit:
fix: pbar should update rather after uploading the file

Commit:
uploads larger files first

Commit:
get_size doesn't use blanket except statement and sets minimum size at 4096

Commit:
fix: exclusion list was not working properly

Commit:
fix: safer way to get file size

Commit:
style: more straightforward message

Commit:
fix: updates the progress bar after each block sent

Commit:
untested: adds rudimentary progress bar when restoring

josephernest commented:
Thank you @thiswillbeyourgithub for this PR. I have been quite busy the last months, so I don't have time to have a look now, but I will definitely look at it when I have time.

Commit:
style: clearer query for user input

Commit:
fix: apparently the progress bar was not updating correctly but the authout might have more luck implementing this

thiswillbeyourgithub commented:
Hi! There is really no rush, I am happy that you don't take it the wrong way as you said PR were not welcomed! But I felt that some features were easy to add and was necessary for me to use it anyway so it was too bad not to share it. Especially a way to exclude some files, have a progress bar and ability to make backups on local drive (especially for the first very large backup).

Commit:
better handling of wrong password

Commit:
minor

Commit:
feat: supports multi threaded uploading

Commit:
updates pbar at each encryption step and not at each uploaded file

Commit:
send small files first by default

Commit:
pbar: change smoothing parameter

thiswillbeyourgithub commented:
Hello again,

I had to use the threading module enough in the summer to feel comfortable enough to implement multithreaded file uploading in my last commits :)

Commit:
better smoothing like that

Commit:
style: shorter print + use f string

Commit:
added a for loop to allow to retry up to 5 times to enter the password

Commit:
style

Commit:
now using tqdm instead of tqdm.tqdm

Commit:
modified:   nfreezer.py
enhanced progress bar when reconstructing distant file list
enhanced progress bar when restoring
added inclusion and exclusion regex

Commit:
oops: forgot to remove a test argument

Commit:
oops: forgot to remove a test argument 2

Commit:
oops: forgot to stop the script

Commit:
adds todo

Commit:
it makes no sense to keep an outdated changelog without timestamp

Commit:
added todo list at the top

Commit:
added SMALL_FILE as a parameter to indicate when to start threading

Commit:
added SMALL_FILE as a parameter to indicate when to start threading 2

Commit:
fix: joining threads at the end

Commit:
modified:   nfreezer.py
* threaded functions are all rewritten, fixed, and moved at the top
* restoring is now also multi threaded and seems to work fine
* numerous other improvement

I am really sorry to put this in such a messy commit but I figured
that with the number of things that want to keep adding it doesn't
make sense anymore to try to follow each chommit and the code change
will have changed quite a lot from the original code

Commit:
todo

Commit:
fix

Commit:
display nunmber of threads in tqdm

Commit:
todo

Commit:
warn user when creating a full backup

Commit:
fix: pbar

Commit:
better way to check if too many threads

Commit:
new defaults

Commit:
added colours to prints

Commit:
minor

Commit:
ask sftp password again if failed
